solutions/cf-660F.mdx

```diff
@@ -0,0 +1,149 @@
+
+$$
+f(r) = \max_{1 \leq l \leq r} \{\texttt{ips}[r] - \texttt{ips}[l - 1] - (l - 1)(\texttt{pre}[r] - \texttt{pre}[l - 1])\}
+$$
+
+$$
+f(r) = \max_{1 \leq l \leq r} \{\texttt{ips}[r] - \texttt{ips}[l - 1] - (l - 1)\texttt{pre}[r] + (l - 1)\texttt{pre}[l - 1]\}
+$$
+
+$$
+f(r) = \max_{1 \leq l \leq r} \{\texttt{ips}[r] - \texttt{ips}[l - 1] - l \cdot \texttt{pre}[r] + \texttt{pre}[r] + l \cdot \texttt{pre}[l - 1] - \texttt{pre}[l - 1]\}
+$$
+
+$$
+f(r) = \max_{1 \leq l \leq r} \{(\texttt{ips}[l - 1] - \texttt{pre}[l - 1] + l \cdot \texttt{pre}[l - 1]) - l \cdot \texttt{pre}[r]\} + \texttt{ips}[r] + \texttt{pre}[r]
+$$
+
+We rearrange the equation such that terms that depend only on $l$ are wrapped in parenthesis inside of the $\max$ statement and terms that depend only on $r$ are moved outside of the $\max$ statement. Note that there is only a single term that depends on both $l$ and $r$. 
+
+Because of this property, we can use convex hull trick. Each value can be represented as a linear function of the form 
+
+$$
+g_l(x) = l \cdot x + (\texttt{ips}[l - 1] - \texttt{pre}[l - 1] + l \cdot \texttt{pre}[l - 1])
+$$
+
+The original equation can thus be expressed as:
+
+$$
+f(r) = \max_{1 \leq l \leq r} \{g_l(\texttt{pre}[r])\} + \texttt{ips}[r] + \texttt{pre}[r]
+$$
+
+**Complexity**: $\mathcal{O}(n)$ or $\mathcal{O}(n \log n)$. The impementation below runs in $\mathcal{O}(n \log n)$ because it uses [Line Container](https://usaco.guide/adv/line-container/). It is possible to get a linear time complexity by taking advantage of the fact that $l$ and $\texttt{pre}[r]$ are both non-decreasing.
+
+# Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+//BeginCodeSnip{Line Container}
+// Source: https://usaco.guide/adv/line-container?lang=cpp
+
+bool _Line_Comp_State; 
+struct Line {
+	// k is slope, m is intercept, p is intersection point
+	mutable long long k, m, p;
+	bool operator<(const Line& o) const {
+		return _Line_Comp_State ? p < o.p : k < o.k;
+	}
+};
+
+struct LineContainer : multiset<Line> {
+	long long div(long long a, long long b) { return a / b - ((a ^ b) < 0 && a % b); }
+
+	bool isect(iterator x, iterator y) {
+		if (y == end()) { 
+			x->p = LLONG_MAX; 
+			return false; 
+		}
+		if (x->k == y->k) 
+			x->p = x->m > y->m ? LLONG_MAX : -LLONG_MAX;
+		else 
+			x->p = div(y->m - x->m, x->k - y->k);
+		return x->p >= y->p;
+	}
+
+	void add(long long k, long long m) {
+		auto z = insert({k, m, 0}), y = z++, x = y;
+		while (isect(y, z)) 
+			z = erase(z);
+		if (x != begin() && isect(--x, y)) 
+			isect(x, y = erase(y));
+		while ((y = x) != begin() && (--x)->p >= y->p)
+			isect(x, erase(y));
+	}
+
+	long long query(long long x) {
+		assert(!empty());
+		_Line_Comp_State = 1; 
+		auto l = *lower_bound({0, 0, x}); 
+		_Line_Comp_State = 0;
+		return l.k * x + l.m;
+	}
+};
+//EndCodeSnip
+
+int N;
+long long A[200005], pre[200005], ips[200005];
```

Comment:
1. put in constant
2. make 0-indexed if possible

Reply:
not necessary, complicates code because the `ips` array is defined as `A[i] * i` (also I use prefix sums so it means an extra `if` condition)

Reply:
well at least put it in a constant of 2e5 and use + 1 or something

Reply:
I did that already

Reply:
well you chanaged it back to 2e5 + 5, good job

Reply:
fixed, pls approve